File: src/Platform/CrazyGames.js
let sdk = null;

export async function init(onMute) {
  const cg = window.CrazyGames?.SDK;
  if (!cg) return;
  try {
    await cg.init();
    sdk = cg;
    sdk.game.addSettingsChangeListener((s) => onMute(!!s.muteAudio));
    onMute(!!sdk.game.settings?.muteAudio);
  } catch (e) {
    console.error("CrazyGames SDK init failed", e);
    sdk = null;
  }
}

const call = (f) => { try { f(); } catch (e) { console.error(e); } };

export function gameplayStart() { if (sdk) call(() => sdk.game.gameplayStart()); }
export function gameplayStop() { if (sdk) call(() => sdk.game.gameplayStop()); }

export function getInviteRoom() {
  if (sdk) {
    try { const r = sdk.game.getInviteParam("roomId"); if (r) return r; } catch (e) { console.error(e); }
  }
  return new URLSearchParams(window.location.search).get("room") ?? "";
}

export function updateRoom(roomId, isJoinable) {
  if (sdk) call(() => sdk.game.updateRoom({ roomId, isJoinable, inviteParams: { roomId } }));
}

export function addJoinRoomListener(callback) {
  if (sdk) call(() => sdk.game.addJoinRoomListener((p) => p?.roomId && callback(p.roomId)));
}

export function isInstantMultiplayer() { return !!sdk?.game?.isInstantMultiplayer; }

export function leftRoom() { if (sdk) call(() => sdk.game.leftRoom()); }

export function isUserAvailable() { return !!sdk?.user?.isUserAccountAvailable; }

export async function getUser() {
  if (!isUserAvailable()) return null;
  try { return await sdk.user.getUser(); } catch (e) { console.error(e); return null; }
}

export async function showAuthPrompt() {
  if (!isUserAvailable()) return null;
  try { return await sdk.user.showAuthPrompt(); } catch (e) { console.error(e); return null; }
}

export function addAuthListener(callback) {
  if (isUserAvailable()) call(() => sdk.user.addAuthListener((u) => u && callback(u)));
}

export function requestAd(adType, adStarted, adFinished, adError) {
  if (!sdk) return adError("no-sdk");
  try {
    sdk.ad.requestAd(adType, { adStarted, adFinished, adError: (e) => adError(String(e?.code ?? e)) });
  } catch (e) {
    adError(String(e));
  }
}

export function dataSetItem(key, value) {
  try {
    (sdk?.data ?? localStorage).setItem(key, value);
  } catch (e) {
    console.error(e);
    call(() => localStorage.setItem(key, value));
  }
}

export function dataGetItem(key) {
  try {
    return (sdk?.data ?? localStorage).getItem(key);
  } catch (e) {
    console.error(e);
    try { return localStorage.getItem(key); } catch (e2) { console.error(e2); return null; }
  }
}

File: src/Platform/CrazyGames.fs
module CrazyGames

open Fable.Core

type CGUser = { username: string; profilePictureUrl: string }

/// True only in the package `scripts/pack-crazygames.mjs` builds, which stamps
/// `window.NDA_CG`; `?cg=1` turns it on locally. The SDK exposes no environment
/// property, so the packer is the signal.
[<Import("isCrazyGames", "./CrazyGames.js")>]
let isCrazyGames () : bool = jsNative

[<Import("init", "./CrazyGames.js")>]
let init (onMute: bool -> unit) : JS.Promise<unit> = jsNative

[<Import("gameplayStart", "./CrazyGames.js")>]
let gameplayStart () : unit = jsNative

[<Import("gameplayStop", "./CrazyGames.js")>]
let gameplayStop () : unit = jsNative

[<Import("loadingStart", "./CrazyGames.js")>]
let loadingStart () : unit = jsNative

[<Import("loadingStop", "./CrazyGames.js")>]
let loadingStop () : unit = jsNative

[<Import("getInviteRoom", "./CrazyGames.js")>]
let getInviteRoom () : string = jsNative

[<Import("updateRoom", "./CrazyGames.js")>]
let updateRoom (roomId: string, isJoinable: bool) : unit = jsNative

/// The room's CrazyGames invite link, fetched by `updateRoom` and cached there;
/// "" until the SDK answers, or on a build with no SDK.
[<Import("inviteLink", "./CrazyGames.js")>]
let inviteLink () : string = jsNative

/// True while the player has chat switched off in the CrazyGames settings.
[<Import("chatDisabled", "./CrazyGames.js")>]
let chatDisabled () : bool = jsNative

[<Import("addJoinRoomListener", "./CrazyGames.js")>]
let addJoinRoomListener (callback: string -> unit) : unit = jsNative

[<Import("isInstantMultiplayer", "./CrazyGames.js")>]
let isInstantMultiplayer () : bool = jsNative

[<Import("leftRoom", "./CrazyGames.js")>]
let leftRoom () : unit = jsNative

[<Import("isUserAvailable", "./CrazyGames.js")>]
let isUserAvailable () : bool = jsNative

[<Import("getUser", "./CrazyGames.js")>]
let getUser () : JS.Promise<CGUser> = jsNative

[<Import("showAuthPrompt", "./CrazyGames.js")>]
let showAuthPrompt () : JS.Promise<CGUser> = jsNative

[<Import("addAuthListener", "./CrazyGames.js")>]
let addAuthListener (callback: CGUser -> unit) : unit = jsNative

[<Import("requestAd", "./CrazyGames.js")>]
let requestAd (adType: string, adStarted: unit -> unit, adFinished: unit -> unit, adError: string -> unit) : unit = jsNative

[<Import("dataSetItem", "./CrazyGames.js")>]
let dataSetItem (key: string, value: string) : unit = jsNative

[<Import("dataGetItem", "./CrazyGames.js")>]
let dataGetItem (key: string) : string = jsNative
